prince/solvers/propagation.py

```diff
@@ -369,194 +378,112 @@ def semi_lagrangian(self, delta_z, z, state):
 
         return state
 
-    def eqn_deriv(self, z, state, *args):
+    def eqn_deriv_standard(self, z, state, *args):
         z = z - self.z_offset
 
         self.ncallsf += 1
 
-        # print 'state', state.shape
         # Update rates/cross sections only if solver requests to do so
         if abs(z - self.current_z_rates) > self.recomp_z_threshold:
             self._update_jacobian(z)
             self.current_z_rates = z
+        
+        if self.using_cupy:
+            import cupy
+            state = cupy.array(state)
 
         r = self.jacobian.dot(state)
-        # print 'deriv', r.shape
-
+        
         if self.enable_injection_jacobian:
-            # print 'inj', self.injection(1., z).shape
-            r += self.injection(1., z)[:,np.newaxis] 
+            r += self.injection(1., z)[:, np.newaxis]
+
         if self.enable_partial_diff_jacobian:
             conloss = np.zeros_like(self.adia_loss_rates_grid.energy_vector)
             if self.enable_adiabatic_losses:
                 conloss += self.adia_loss_rates_grid.loss_vector(z)
             if self.enable_pairprod_losses:
                 conloss += self.pair_loss_rates_grid.loss_vector(z)
-            # print 'conloss', conloss.shape
-            # print 'state',state.shape
-            partial_deriv = self.dldz(z) * self.diff_operator.operator.dot(
-                conloss[:,np.newaxis] * state)
-            # print 'pderiv', partial_deriv.shape
-            r += partial_deriv
-
+            r += self.dldz(z) * self.diff_operator.dot(
+                conloss[:, np.newaxis] * state)
         return r
 
-    def eqn_deriv_euler(self, z, state, *args):
```

Comment:
I don't even remember that it worked and the status is. If it works, then speed gains will be serious for MKL/CUDA because of no LU. The problem with the current implementation is that the step size is not adaptive and to maintain stability we need to know the leading eigenvalues which should be for our matrix dldz*max(diag(coupling_mat)) if the step is fast, then Euler may kick a**. I'll try this out.